From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: slothfulchat <patches@slothfulchat.invalid>
Date: Wed, 14 Jan 2026 18:59:28 +0100
Subject: [PATCH 25/33] fix: add `ChatlistItemChanged` event to
 `on_archived_chats_maybe_noticed`

Also fixes duplicated events when marking archived link as noticed.
---
 src/chat.rs | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/src/chat.rs b/src/chat.rs
index f2d4080..58ee0b3 100644
--- a/src/chat.rs
+++ b/src/chat.rs
@@ -3297,7 +3297,10 @@ pub async fn marknoticed_chat(context: &Context, chat_id: ChatId) -> Result<()>
 
     context.emit_event(EventType::MsgsNoticed(chat_id));
     chatlist_events::emit_chatlist_item_changed(context, chat_id);
-    context.on_archived_chats_maybe_noticed();
+    if !chat_id.is_archived_link() {
+        // prevents event duplication when marking all archived chats as noticed
+        context.on_archived_chats_maybe_noticed();
+    }
     Ok(())
 }
 
@@ -5318,12 +5321,14 @@ pub(crate) async fn sync_alter_chat(&self, id: &SyncId, action: &SyncAction) ->
         }
     }
 
-    /// Emits the appropriate `MsgsChanged` event. Should be called if the number of unnoticed
+    /// Emits the appropriate `MsgsChanged` and `ChatlistItemChanged` event.
+    /// Should be called if the number of unnoticed
     /// archived chats could decrease. In general we don't want to make an extra db query to know if
     /// a noticed chat is archived. Emitting events should be cheap, a false-positive `MsgsChanged`
     /// is ok.
     pub(crate) fn on_archived_chats_maybe_noticed(&self) {
         self.emit_msgs_changed_without_msg_id(DC_CHAT_ID_ARCHIVED_LINK);
+        chatlist_events::emit_chatlist_item_changed(self, DC_CHAT_ID_ARCHIVED_LINK);
     }
 }
 
